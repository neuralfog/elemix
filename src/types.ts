import { type HtmlTemplate, html as rhtml } from '@neuralfog/elemix-renderer';

export type Template = HtmlTemplate;
export const html = rhtml;

export const RenderTrigger = {
    PROPS: 'PROPS',
    SIGNAL: 'SIGNAL',
    LOCAL_STATE: 'LOCAL_STATE',
    ON_MOUNT: 'ON_MOUNT',
} as const;

export type RenderTriggerType =
    | (typeof RenderTrigger)[keyof typeof RenderTrigger]
    | string;
