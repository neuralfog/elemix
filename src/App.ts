type AppConfig = {
    cssReset?: string;
    entryPoint?: () => Promise<any>;
};

// biome-ignore lint:
export class App {
    static config: AppConfig = {};

    public static mergeConfigs(config: AppConfig): void {
        App.config = {
            ...App.config,
            ...config,
        };
    }
}

export const initApp = (config: AppConfig): void => {
    App.mergeConfigs(config);
    void App.config.entryPoint?.();
};
