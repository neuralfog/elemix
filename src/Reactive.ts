import type { Component } from './component/Component';
import type { RenderTriggerType } from './types';

export class Reactive<State> {
    public subscribers = new Set<Component>();
    private proxy: State;
    private proxySet = new WeakSet<any>();

    public get value(): State {
        return this.proxy;
    }

    constructor(
        state: any,
        private renderTrigger?: RenderTriggerType,
    ) {
        this.proxy = this.create(state);
    }

    public subscribe(instance: Component): Reactive<State> {
        if (!this.subscribers.has(instance)) this.subscribers.add(instance);
        return this;
    }

    public unsubscribe(instance: Component): Reactive<State> {
        this.subscribers.delete(instance);
        return this;
    }

    private create(state: any): State {
        const reactive = this;

        return new Proxy(state, {
            get(target, name): unknown {
                const prop = target[name];

                unsupportedCollectionError(prop);

                if (
                    typeof prop === 'object' &&
                    prop !== null &&
                    !reactive.proxySet.has(prop)
                ) {
                    const propRef = new Proxy(prop, this);
                    reactive.proxySet.add(propRef);
                    return propRef;
                }

                return prop;
            },
            set(target, name, value): boolean {
                target[name] = value;

                unsupportedCollectionError(target);

                if (Array.isArray(target) && name === 'length') {
                    reactive.notify();
                    return true;
                }

                reactive.notify();
                return true;
            },
        }) as State;
    }

    private notify(): void {
        for (const subscriber of this.subscribers) {
            subscriber.render(this.renderTrigger);
        }
    }
}

export const UNSUPPORTED_COLLECTION_ERROR_MESSAGE =
    'Reactive state does not support collections: Map, WeakMap, Set, WeakSet';

const unsupportedCollectionError = (prop: any): any => {
    if (
        prop instanceof Map ||
        prop instanceof WeakMap ||
        prop instanceof Set ||
        prop instanceof WeakSet
    ) {
        throw new Error(UNSUPPORTED_COLLECTION_ERROR_MESSAGE);
    }
};
