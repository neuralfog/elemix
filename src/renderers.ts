import type { Renderer } from './component/Renderer';

export const activeRenderers = new Set<Renderer>();
