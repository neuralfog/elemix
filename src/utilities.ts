export const ref = <Value>(
    value: Value | undefined = undefined,
): { value?: Value } => {
    return {
        value,
    };
};
