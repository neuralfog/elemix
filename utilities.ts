export { mergeClasses } from '@neuralfog/elemix-renderer/utils';
export * from './src/utilities';
