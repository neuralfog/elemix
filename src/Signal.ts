import { Reactive } from './Reactive';
import { RenderTrigger } from './types';

export const signal = <State>(
    state: State,
    renderTrigger?: string,
): Reactive<State> =>
    new Reactive(state, renderTrigger || RenderTrigger.SIGNAL);
