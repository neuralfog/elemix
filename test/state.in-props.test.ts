import { expect, test, describe, beforeEach, vi } from 'vitest';
import { HTML } from '@neuralfog/elemix-testing/snapshots';
import { html } from '@neuralfog/elemix-renderer';
import { Present } from '@neuralfog/elemix-testing';
import type { StateInProps, StateInPropsChild } from './fixtures/StateInProps';
import { RenderTrigger } from '../src/types';
import { render } from '../utilities';

import './fixtures/StateInProps';

describe('State In Props', () => {
    beforeEach(() => {
        document.body.innerHTML = '';
    });

    test('Initial State', async () => {
        const presenter = new Present().screen(html`<state-in-props />`);

        await render();

        expect(HTML(presenter.root<StateInProps>())).toMatchSnapshot();
    });

    test('Change Value State In Props', async () => {
        const presenter = new Present().screen(html`<state-in-props />`);
        const parent = presenter.root<StateInProps>();

        await render();

        const child = presenter.getByTag<StateInPropsChild>(
            'state-in-props-child',
        );

        await render();

        const onRenderParent = vi.spyOn(parent, 'onRender');
        const onRenderChild = vi.spyOn(child, 'onRender');

        child.props.state.nestedValue = 'Changed Value From Child Level';

        await render();

        expect(onRenderParent).toHaveBeenCalledOnce();
        expect(onRenderParent).toHaveBeenCalledWith([
            RenderTrigger.LOCAL_STATE,
        ]);
        expect(onRenderChild).toHaveBeenCalledOnce();
        expect(onRenderChild).toHaveBeenCalledWith([RenderTrigger.PROPS]);

        expect(HTML(presenter.root<StateInProps>())).toMatchSnapshot();
    });
});
