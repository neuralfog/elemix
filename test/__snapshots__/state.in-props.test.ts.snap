// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`State In Props > Change Value State In Props 1`] = `
"
<state-in-props><!-- #shadowroot -->
  <h1>StateInProps</h1>
  <p>Changed Value From Child Level</p>
  <state-in-props-child><!-- #shadowroot -->
    <h1>StateInPropsChild</h1>
  </state-in-props-child>
</state-in-props>
"
`;

exports[`State In Props > Initial State 1`] = `
"
<state-in-props><!-- #shadowroot -->
  <h1>StateInProps</h1>
  <p>Nested Initial Value</p>
  <state-in-props-child><!-- #shadowroot -->
    <h1>StateInPropsChild</h1>
  </state-in-props-child>
</state-in-props>
"
`;
